### ISSUE TITLE
failed to load nvidia module - device not present - incorrect IRONHIDE_ACPI_MODE for XPS 15 L502x

### ISSUE THREAD

**Issue by rockorequin** [MrMEEE/ironhide#84]
I get the 'failed to start ironhide server' problem on my XPS 15 (L502z), and it's because the nvidia-current module won't load, saying device not present.

The bug is an old one that I recall from the bumblebee days: the card is being enabled then disabled at startup via the ironhide-disablecard-on-powerup, which confuses the nvidia module. Looking at the  ironhide-disablecard-on-powerup script, it is because the configuration in /etc/default/ironhide has:

```
IRONHIDE_ACPI_MODE=1
```

Who sets this? I don't recall seeing a question about it during install. Can we change the configuration database for the L502x so it doesn't get set in the future?

Another thought: might the problem occur because the card is enabled and then disabled too quickly, before the nvidia module has a chance to complete initialising? After all, once in an X session, I can manually issue the commands to enable and disable the card and that doesn't bork the nvidia module.

**Comment by graydw57** [comment]
I'm having the same issue. Something has changed - this worked for me initially, but now I get the same
device not present error. According to the kernel log, this card is not supported by this version of the 
nvidia driver (280.13 from X-swat/updates). Huh?
Did Nvidia remove support?

**Comment by rockorequin** [comment]
Have you tried setting IRONHIDE_ACPI_MODE=0 in /etc/default/ironhide and then rebooting?

**Comment by graydw57** [comment]
Ah, sorry, been away for a while. 

Did some experimenting - removing ironhide (ppa_purge) and reinstalling - worked! OK, reboot. Doesn't work. Change the
IRONHIDE_ACPI_MODE flag to 0, works all the time. 

OT - I had the chance to play with the webgl stuff. Doesn't work on the Intel side, but with the Nvidia card running, it's 
working fine. The Flight of the Navigator demo is beautiful in Chrome (albeit the Flickr links don't work - it may
be a case of bitrot.) 

So, I'm happy.

**Comment by yorikvanhavre** [comment]
I have a similar problem on a Vostro3500 with oneiric... ironhide didn't detect/copy the correct enable/disable scripts, so I did it manually. They seem to turn the card on/off fine, but the nvidia-current kernel module cannot be loaded (Error inserting nvidia-current, no such  device), either by the script or manually with modprobe. Changing IRONHIDE_ACPI_MODE flag to 0 didn't solve... I'm sending a bugreport right now.

Can there be other causes outside the enable/disable issue mentioned in the first post?

**Comment by rockorequin** [comment]
@yorikvanhavre: Just checking, did you reboot after changing the flag?

**Comment by yorikvanhavre** [comment]
yes i did... the difficulty is that I'm not sure what exactly is supposed to happen... Normally I should be able to turn it off then on again manually, then modprobe the driver, right?

**Comment by rockorequin** [comment]
ironhide-enablecard and ironhide-disablecard should do that for you, including modprobing and rmmodding it (you get no error message if it works). Or running 'optirun glxspheres' should test if it is working.

I suppose the next thing is to check that you have the correct enable ACPI script for the Vostro.

**Comment by yorikvanhavre** [comment]
Thanks for the help rockorequin...

I believe I have the correct script... The acpi_call module is already included in ubuntu, but I downloaded the source code of it, which comes with a test script... The test script output is this:

`$ ./test_off.sh 
Trying \_SB.PCI0.P0P1.VGA._OFF: failed
Trying \_SB.PCI0.P0P2.VGA._OFF: failed
Trying \_SB_.PCI0.OVGA.ATPX: failed
Trying \_SB_.PCI0.OVGA.XTPX: failed
Trying \_SB.PCI0.P0P3.PEGP._OFF: failed
Trying \_SB.PCI0.P0P2.PEGP._OFF: failed
Trying \_SB.PCI0.P0P1.PEGP._OFF: works!`

The last one is the same that is used in my ironhide enable/disable scripts. Either after enabling or disabling, still no way to modprobe the nvidia module. Just a dumb question, the card must be enabled before inserting the nvidia module, right?

Another question, do you know how to check if the card is actually enabled or disabled? I did lcpci -vnnn | grep VGA, which shows a different text if the card is (supposedly) on or off, but it doesn't tell explicitely...

**Comment by rockorequin** [comment]
I get the following output from lspci -vnnn|grep VGA:
1. When the card is OFF:
   
   ```
   01:00.0 VGA compatible controller [0300]: nVidia Corporation GF106 [GeForce GT 555M] [10de:0df4] (rev ff) (prog-if ff)
   ```
2. When the card is ON:
   
   ```
   01:00.0 VGA compatible controller [0300]: nVidia Corporation GF106 [GeForce GT 555M] [10de:0df4] (rev a1) (prog-if 00 [VGA controller])
   ```

So the difference is that it says "prog-if 00 [VGA controller]" if it is on and just "prog-if ff" if it is off.

So how did you figure out the command to turn it on? I guessed the obvious and replaced _OFF with _ON in the enable script, and it worked, but maybe it is different on the Vostro?

**Comment by yorikvanhavre** [comment]
Okay, thanks a lot for testing! I get the exact same output as you... So I guess that part works fine.
Yes, I just replaced _OFF by _ON, which seems indeed to be the right way.
But I found some more weird info from people with the same notebook in this other thread:
http://github.com/MrMEEE/bumblebee/issues/171
Looks like the nvidia driver gets confused with all the turn off/turn on... Seems specific to some hardware... I'll try the things they suggest there (and report here, in case others read this).

**Comment by yorikvanhavre** [comment]
Wow, there was an upgrade yesterday don't know what it was (ironhide or nvidia driver, probably ironhide) but now everything magically works! The enable/disable scripts are now totally empty, they don't even contain the modprobe command anymore (but the nvidia driver is well loaded at boot)

So I don't have anymore problems! MrEEE for president!
